fpga_interchange/examples/tests.cmake

```diff
@@ -0,0 +1,125 @@
+function(add_interchange_test)
+    # ~~~
+    # add_interchange_test(
+    #    name <name>
+    #    device <common device>
+    #    package <package>
+    #    tcl <tcl>
+    #    xdc <xdc>
+    #    top <top name>
+    #    sources <sources list>
+    #    [techmap <techmap file>]
+    # )
+    #
+    # Generates targets to run desired tests
+    #
+    # Targets generated:
+    #   - test-fpga_interchange-<name>-json     : synthesis output
+    #   - test-fpga_interchange-<name>-netlist  : interchange logical netlist
+    #   - test-fpga_interchange-<name>-phys     : interchange physical netlist
+    #   - test-fpga_interchange-<name>-dcp     : design checkpoint with RapidWright
+
+    set(options)
+    set(oneValueArgs name device package tcl xdc top techmap)
+    set(multiValueArgs sources)
+
+    cmake_parse_arguments(
+        add_interchange_test
+        "${options}"
+        "${oneValueArgs}"
+        "${multiValueArgs}"
+        ${ARGN}
+    )
+
+    set(name ${add_interchange_test_name})
+    set(device ${add_interchange_test_device})
+    set(package ${add_interchange_test_package})
+    set(top ${add_interchange_test_top})
+    set(tcl ${CMAKE_CURRENT_SOURCE_DIR}/${add_interchange_test_tcl})
+    set(xdc ${CMAKE_CURRENT_SOURCE_DIR}/${add_interchange_test_xdc})
+    set(techmap ${CMAKE_CURRENT_SOURCE_DIR}/${add_interchange_test_techmap})
+
+    set(sources)
+    foreach(source ${add_interchange_test_sources})
+        list(APPEND sources ${CMAKE_CURRENT_SOURCE_DIR}/${source})
+    endforeach()
+
+    if (NOT DEFINED top)
+        # Setting default top value
+        set(top "top")
+    endif()
+
+    # Synthesis
+    set(synth_json ${CMAKE_CURRENT_BINARY_DIR}/${name}.json)
+    add_custom_command(
+        OUTPUT ${synth_json}
+        COMMAND
+            SOURCES=${sources}
+            OUT_JSON=${synth_json}
+            TECHMAP=${techmap}
+            yosys -c ${tcl}
+        DEPENDS ${sources}
+    )
+
+    add_custom_target(test-${family}-${name}-json DEPENDS ${synth_json})
```

Comment:
Where did family variable come from?

Reply:
In general no variables should be used like this unless they come from an argument to the function.

Reply:
Ah, `family` was from the `nextpnr` root. Please add a comment to that affect.

Reply:
I'd prefer we pass the family variable in.  See basically add a `oneValueArg`, `family`.

And then when invoking `add_interchange_test` add `family ${family}`.

Reply:
Given that the family will alwasy be fpga_interchange (if we can assume this), might be worth avoiding adding additional arguments and have targets like `test-fpga-interchange-${name}-json`

Reply:
In the long run, we are probably going to want other roll up targets.  For example, all A7 tests, all Z7 tests, all Nexus tests, all Zynq US+ tests, etc.  If you make the `family` argument explicit, there is an easier attachment point.

Reply:
All right, done